### ISSUE TITLE
openPMD HDF5 output takes too long - ADIOS2 or Plotfiles faster

### ISSUE THREAD

**Issue by kookjine** [ECP-WarpX/WarpX#3227]
Hello,

My simulation is taking approximately 30 minutes for saving data at every saving step for numerical parameters as below.
```
Initializing CUDA...
CUDA initialized with 1 GPU per MPI rank; 16 GPU(s) used in total
MPI initialized with 16 MPI processes
MPI initialized with thread support level 3
AMReX (22.07-4-gcbdc6580ee3d) initialized
WarpX (22.06-33-g590b4bedaa8a)
PICSAR (4252e567089f)
Level 0: dt = 2.005586005e-14 ; dx = 8.59375e-06 ; dy = 8.59375e-06 ; dz = 4.150390625e-05

Grids Summary:
Level 0 16 grids 67108864 cells 100 % of domain
smallest grid: 256 x 128 x 128 biggest grid: 256 x 128 x 128

--------------------------- MAIN EM PIC PARAMETERS ----------------------------
Precision: | DOUBLE
Particle precision: | DOUBLE
Geometry: | 3D (XYZ)
Operation mode: | Electromagnetic (vacuum)
Current Deposition: | Esirkepov
Particle Pusher: | Boris
Charge Deposition: | standard
Field Gathering: | energy-conserving
Particle Shape Factor: | 3
Maxwell Solver : | Yee
Moving window: | ON (moving_window_dir = z, moving_window_v = 299792458)

STEP 500 starts ...
--- INFO : re-sorting particles
STEP 500 ends. TIME = 1.002793002e-11 DT = 2.005586005e-14
Evolve time = 1780.529649 s; This step = 1739.380607 s; Avg. per step = 3.561059297 s
```
Now I am not sure if this simulation is working fine..
Is this normal?

++ The WarpX program was compiled by CMake based on GCC 8.5.0, CUDA Toolkit 11.0, OpenMPI 4.1.1, and all other dependencies from Spack.
(To be honest I am not sure if my OpenMPI library was properly built for WarpX. I have had issues of properly building WarpX with the right OpenMPI library.)
Attached is the zip file of WarpX input and Slurm submission scripts.
[Neuron.zip](https://github.com/ECP-WarpX/WarpX/files/9088269/Neuron.zip)
I will be very grateful if I can get any hint of this issue.

Thanks a lot!

**Comment by guj** [comment]
It is not normal!  Which machine is it on?

Would you try again and see if it repeats.

Thanks,
- Junmin

**Comment by kookjine** [comment]
It's on a local machine with 4 nodes, which each node includes 4 TITAN V GPUS.
It's repeating same issue.
Thanks!
Kook-Jin

**Comment by ax3l** [comment]
As Junmin already wrote, this is not normal.
Do you use HDF5 or ADIOS2 for openPMD output?

Where do you get the OpenMPI from? In case OpenMPI was not built with Spack, then you might want to register it as an external package in Spack:
https://spack.readthedocs.io/en/latest/build_settings.html

**Comment by kookjine** [comment]
I usually use hdf5 format for OpenPMD output.
The openpmd-api was built with Spack.
GCC, CUDA, and OpenMPI were built from source code.
GCC and OpenMPI were registered in ~/.spack/linux/compilers.yaml and ~/.spack/packages.yaml, respectively. 
I didn't check how ADIOS2 works yet.

**Comment by guj** [comment]
Since you have all software in Spack, would you change the line in input:
diag1.openpmd_backend = h5
to
diag1.openpmd_backend = bp

and run again.  Please let me know how it works.  How big is your file? And
what file system do you have?

- Junmin

**Comment by kookjine** [comment]
The file system is shown to be 'xfs' and each file size I got from the simulation is ~29 G.
I saw in another test script with 'plotfile' format, 5 G data files were saved in 5 s..
Maybe it is about data format type as you expect. 
I will let you know how the attached script works with 'openpmd.bp' and 'plotfile' when I got slots in the cluster.
Thank you!
Kook-Jin

**Comment by guj** [comment]
Yes, would be good to test out plot file as well.

Thanks

**Comment by kookjine** [comment]
Saving data file took ~11 s for bp, ~7 s for plotfile format with attached scripts.
So, it seems that the issue is happening only with h5 format.
Thanks,
Kook-Jin

**Comment by ax3l** [comment]
Hi @kookjine,

I fear that's a common behavior that HDF5 needs extra tuning to perform efficiently. That's exactly why we implement multiple backends in openPMD, such as ADIOS2.

Do you cluster admins document any (environment) hints to tune parallel HDF5 performance on the system? What's the HDF5 version you are using?
Otherwise just stay with ADIOS2 `.bp` files or plotfiles, that's why we implement them to be plug-and-play :)

Best,
Axel

**Comment by ax3l** [comment]
I am cc-ing @jeanbez @sbyna in case they are curious what's wrong with HDF5's performance on your system. I am not sure how much work goes in tuning HDF5 on xfs.

**Comment by kookjine** [comment]
Hi Axel @ax3l 

Thank you for caring for the issue.
I just didn't know I can use ADIOS2 same way as HDF5 in OpenPMD.
Now for me working with ADIOS2.bp is totally fine.

By the way, the issue does not happen when I use h5 format of OpenPMD in FBPIC or WARP in the same cluster.
Usually I get HDF5 from the package managers recommended by developers such as conda or spack.
The current HDF5 version for WarpX is 1.10.7.

Thanks a lot,
Kook-Jin

**Comment by guj** [comment]
Thanks for getting back the numbers!

Yes, most likely it is a hdf5 issue.  I have not used xfs before.  I know
on some other systems increasing the I/O block size on the file system can
improve hdf5 performance.

On the other hand, is it possible for you to use a non hdf5 format?

- Junmin

**Comment by jeanbez** [comment]
@kookjine I believe you need a newer version of HDF5, at least 1.12 would be ideal. @ax3l are you pining this to 1.10.7 for any particular reason?

**Comment by ax3l** [comment]
Thanks @jeanbez,

>  The current HDF5 version for WarpX is 1.10.7.

Hm, I think I am not... let me check. *update:* ah yes, the latest 1.12 release on Summit has the performance bug that you found.

@kookjine are you on Summit? Looks like we can go now up to 1.12.1
```
$ module av 2>&1 | grep -i hdf5
   hdf5/1.8.22                            slepc/3.15.0
   hdf5/1.10.6                            slepc/3.15.1
   hdf5/1.10.7                   (L,D)    slepc/3.16.0-cpu
   hdf5/1.12.1                            slepc/3.16.0
   hdf5/1.10.7                          superlu/5.2.2
   hdf5/1.10.7
   hdf5_perf/1.10.6.gcc             pwanalyzer/1.3.1
   hdf5_perf/1.10.6.xl       (D)    pwanalyzer/1.4.1              (D)
```

**Comment by ax3l** [comment]
Updating the docs in https://github.com/ECP-WarpX/WarpX/pull/3838 - this is the only location I found for this specific version :)

**Comment by ax3l** [comment]
@jeanbez doesn't 1.12.1 have the meta-data issue you discovered & fixed?
I will ask for a newer HDF5 module on Summit via ticket...

https://openpmd-api.readthedocs.io/en/0.15.1/backends/hdf5.html#known-issues

**Comment by ax3l** [comment]
> ++ The WarpX program was compiled by CMake based on GCC 8.5.0, CUDA Toolkit 11.0, OpenMPI 4.1.1, and all other dependencies from Spack.

Ah, @kookjine actually installed via Spack.
The approach to get a newer HDF5 here is to update the Spack installation _and_ environment. Spack already ships `1.14.0`. I assume Spack pinned the older version at the time the issue was opened...

**Comment by jeanbez** [comment]
> @jeanbez doesn't 1.12.1 have the meta-data issue you discovered & fixed? I will ask for a newer HDF5 module on Summit via ticket...
> 
> https://openpmd-api.readthedocs.io/en/0.15.1/backends/hdf5.html#known-issues

No, it was fixed first in 1.12
